### ISSUE TITLE
Under the goss mode, the l2 loss is suddenly increased during training

### ISSUE THREAD

**Issue by shaoguang123** [microsoft/LightGBM#3053]
## How you are using LightGBM?

LightGBM component: **Python package**

## Environment info

Operating System: **Windows**

CPU/GPU model: **CPU**

Other:

LightGBM version or commit hash:   **2.3.1**

## The log info and loss curve is below：
![log](https://user-images.githubusercontent.com/7996409/81143075-bfe33080-8fa3-11ea-90a5-d8acdd58612d.png)

## The parameters：
`model = LGBMRegressor(max_depth=10, learning_rate=0.01, n_estimators=1000, num_leaves=250,min_child_samples=200, colsample_bytree=0.2, boosting_type="goss", top_rate=0.95, other_rate=0.01, random_state=0, n_jobs=52)`

## data info:
the dataset is about 1300000 samples and 3500 features

## my questions

1. how did this happen?
In my understanding, the prediction is the sum of the leaves values which times learning_rate. And the leaf's value is calculated by 
![w](https://user-images.githubusercontent.com/7996409/81144617-0ab27780-8fa7-11ea-81c1-eb5f64f0721e.png)
The loss function I used is mse, so H is 2, G is the residual.

2. how can I prevent it 

## Similar case in Kaggle
https://www.kaggle.com/c/home-credit-default-risk/discussion/60921

**Comment by guolinke** [comment]
I think it is caused by the numerical issues. 
can you try the fix in https://github.com/microsoft/LightGBM/tree/goss-fix?

**Comment by guolinke** [comment]
BTW, you can record/output the `multiply` here: https://github.com/microsoft/LightGBM/commit/70e1ad6e9c6dd09f0e351c96cfbd88db06e25bb1#diff-30ab9be51da7b46192ac5c2de6afba0aR92, if it is too small or large, it will cause the problem.

**Comment by shaoguang123** [comment]
Thanks a lot for your reply! 
I tried the goss-fix version, but I couldn't output the multipy. 
I just add the code as below:
`score_t multiply = std::max(kEpsilon, static_cast<score_t>(cnt - top_k) / other_k);
Log::Info("multiply: %f", multiply);`
I don't know why this didn't work. 

BTW,In my regression task, the label is between -0.5 and 0.5.

**Comment by shaoguang123** [comment]
Is there something wrong with my log code? Or the code not be executed？

**Comment by shaoguang123** [comment]
ohh, I find the reason why the log not outout.
In the version  `goss-fix` you gave me, the function `BaggingHelper` in goss.h is not be executed, because the function `GBDT::Bagging(iter)`  don't work under goss mode.

**Comment by shaoguang123** [comment]
> ohh, I find the reason why the log not outout.
> In the version `goss-fix` you gave me, the function `BaggingHelper` in goss.h is not be executed, because the function `GBDT::Bagging(iter)` don't work under goss mode.

Actually, the version goss-fix doesn't  implement the effect of `goss`.

**Comment by guolinke** [comment]
@shaoguang123 I think it is a bug introduced by the last refactoring. can you try it again?

**Comment by shaoguang123** [comment]
> @shaoguang123 I think it is a bug introduced by the last refactoring. can you try it again?

Of course，Thank you again for your prompt reply. I will post the result after the test.

**Comment by shaoguang123** [comment]
> @shaoguang123 I think it is a bug introduced by the last refactoring. can you try it again?

This code is wrong
![微信图片_20200508154022](https://user-images.githubusercontent.com/7996409/81383224-4170c500-9142-11ea-96db-27ad461f946c.png)
@guolinke

**Comment by shaoguang123** [comment]
I updated the `goss.h` file the same as you in `version 2.3.1` in order to run the model to get the log output.

However, all the three checks has no error output. and there is no exception about the multiply.

**Comment by guolinke** [comment]
sorry, my bad. 
can you try to use `max_delta_step`, https://lightgbm.readthedocs.io/en/latest/Parameters.html#max_delta_step, to limit the max leaf output?

**Comment by shaoguang123** [comment]
After trying to set `max_delta_step` = 0.5/0.1/0.05 (`learning_rate` = 0.01), the l2 loss doesn't become so large as before.

max_delta_step = 0.5
![0 5](https://user-images.githubusercontent.com/7996409/81468212-c2968d80-9210-11ea-9469-6b6ef7382726.png)

max_delta_step = 0.1
![0 1](https://user-images.githubusercontent.com/7996409/81468224-cde9b900-9210-11ea-9eb3-5b441f59cdba.png)

max_delta_step = 0.05
![0 05](https://user-images.githubusercontent.com/7996409/81468234-e2c64c80-9210-11ea-9100-5e2cbae114d7.png)

But I still don't understand why I encounter the convergence issue with 'goss', and what makes the convergence of 'goss' different from 'gbdt'?

**Comment by guolinke** [comment]
if `max_delta_step` works, it means there are very leaf outputs in the training.
there are many possibilities, for example, some outliers, with large gradients, are sampled, and we give a large weight to them. 

Maybe you can check the leaf outputs in the model file, and found out when and how it happens.

**Comment by shaoguang123** [comment]
> if `max_delta_step` works, it means there are very leaf outputs in the training.
> there are many possibilities, for example, some outliers, with large gradients, are sampled, and we give a large weight to them.
> 
> Maybe you can check the leaf outputs in the model file, and found out when and how it happens.

why we give a large weight to the outliers with large gradients?
Aren't we just give the small gradients a constant weight determined by top_rate and other_rate.

**Comment by shaoguang123** [comment]
> > if `max_delta_step` works, it means there are very leaf outputs in the training.
> > there are many possibilities, for example, some outliers, with large gradients, are sampled, and we give a large weight to them.
> > Maybe you can check the leaf outputs in the model file, and found out when and how it happens.
> 
> why we give a large weight to the outliers with large gradients?
> Aren't we just give the small gradients a constant weight determined by top_rate and other_rate.

BTW, The smaller the top_rate and other rate, the more often this happens.

**Comment by shaoguang123** [comment]
I checked the leaf outputs in the model file, and found out the abnormal tree as below.
![tree](https://user-images.githubusercontent.com/7996409/81470974-a26eca80-9220-11ea-9448-88205114d64d.png)

We can see that the internal_value of leaf 1 is abnormally large。

**Comment by shaoguang123** [comment]
> > > if `max_delta_step` works, it means there are very leaf outputs in the training.
> > > there are many possibilities, for example, some outliers, with large gradients, are sampled, and we give a large weight to them.
> > > Maybe you can check the leaf outputs in the model file, and found out when and how it happens.
> > 
> > 
> > why we give a large weight to the outliers with large gradients?
> > Aren't we just give the small gradients a constant weight determined by top_rate and other_rate.
> 
> BTW, The smaller the top_rate and other rate, the more often this happens.

I checked the leaf outputs in the model file, and found out the abnormal tree as below.
![tree](https://user-images.githubusercontent.com/7996409/81470974-a26eca80-9220-11ea-9448-88205114d64d.png)

We can see that the internal_value of leaf 1 is abnormally large。

**Comment by shaoguang123** [comment]
I checked the leaf outputs in the model file, and found out the abnormal tree as below.
![tree](https://user-images.githubusercontent.com/7996409/81471622-fe3b5280-9224-11ea-838c-bcfc97c93905.png)


We can see that the internal_value of leaf 1 is abnormally large。@guolinke
BTW, can you explain the different among ‘internal_value’, ‘internal_count’ and ‘internal_weight’ when ploting the tree. @guolinke

**Comment by guolinke** [comment]
@shaoguang123 you are right. Let us analysis from the model file. 
I cannot see the interval_xxx in your figures. 
`interval_value` is the output of that node, when it was a leaf (but latter split to two leaves).
`interval_count` is the number of data of that node, and `interval_weight` is the sum of hessian.

**Comment by shaoguang123** [comment]
In order to make trees smaller, I run a demo with parameters as blew.
`model = LGBMRegressor(max_depth=6, learning_rate=0.04, n_estimators=150, num_leaves=30,min_child_samples=200, colsample_bytree=0.2, boosting_type="goss", top_rate=0.5, other_rate=0.1, random_state=0, n_jobs=52)`

I find leaf value increased from the 45th tree's leaf 1(surrounded by red lines in the figure `tree_45.png`). And than the error is passed down.

tree_44.png
![tree_44](https://user-images.githubusercontent.com/7996409/81529566-b2f28280-9391-11ea-8a1e-34ec2007232d.png)

tree_45.png
![tree_45](https://user-images.githubusercontent.com/7996409/81529591-c00f7180-9391-11ea-845c-6a7bef088ca6.png)

tree_46.png
![tree_46](https://user-images.githubusercontent.com/7996409/81529620-cef62400-9391-11ea-9555-60904477c646.png)

tree_47.png
![tree_47](https://user-images.githubusercontent.com/7996409/81529645-d9b0b900-9391-11ea-822d-0320a8c7cf9d.png)

tree_48.png
![tree_48](https://user-images.githubusercontent.com/7996409/81529663-e503e480-9391-11ea-85c4-0ef4dacba750.png)

**Comment by shaoguang123** [comment]
Because my learning_rate=0.04, the first 25 trees don't use "goss". And in the first 25 trees, Leaf weight equals Leaf count. As what you said before,  it means the hessian equals 1?

**Comment by shaoguang123** [comment]
The weights of the 45th tree's first branch are weird。
![tree_45_branch](https://user-images.githubusercontent.com/7996409/81533085-24cdca80-9398-11ea-8f77-99c28757d7ab.png)

The upper one has 33% data and the weight is 1259660, while the other one has 66% data and the weight is 12.998.

**Comment by shaoguang123** [comment]
> The weights of the 45th tree's first branch are weird。
> ![tree_45_branch](https://user-images.githubusercontent.com/7996409/81533085-24cdca80-9398-11ea-8f77-99c28757d7ab.png)
> 
> The upper one has 33% data and the weight is 1259660, while the other one has 66% data and the weight is 12.998.

I have run several models and find that the first branch of the first abnormal tree always divided the data into two parts, accounting for 60% and 30% respectively(Like the picture above).

**Comment by guolinke** [comment]
@shaoguang123 did this happen only in `goss` mode?
can you also check the data in `column_1300`, is there any pattern in this feature?
BTW, can you also try the latest master branch? I just fix the goss. maybe there is a bug that already fixed.

**Comment by shaoguang123** [comment]
> @shaoguang123 did this happen only in `goss` mode?
> can you also check the data in `column_1300`, is there any pattern in this feature?
> BTW, can you also try the latest master branch? I just fix the goss. maybe there is a bug that already fixed.

Yes, it only happens in `goss` mode.
I checked the feature column_1300 and the first splited feature of the first abnormal tree in other models , but there is nothing special.
I will try the latest master branch immediately.

**Comment by guolinke** [comment]
@shaoguang123 if master still didn't work, you can try this branch: https://github.com/microsoft/LightGBM/tree/temp-goss

**Comment by shaoguang123** [comment]
> @shaoguang123 if master still didn't work, you can try this branch: https://github.com/microsoft/LightGBM/tree/temp-goss

ok

**Comment by shaoguang123** [comment]
> @shaoguang123 if master still didn't work, you can try this branch: https://github.com/microsoft/LightGBM/tree/temp-goss

ok

**Comment by shaoguang123** [comment]
> > @shaoguang123 if master still didn't work, you can try this branch: https://github.com/microsoft/LightGBM/tree/temp-goss
> 
> the master version was stuck at the 25th iteration as my learning_rate is 0.04.
> when I use the version temp-goss, the log is `[Fatal] Check failed: (best_split_info.left_count) > (0)` @guolinke

at src\treelearner\serial_treelearner.cpp, line 623

**Comment by guolinke** [comment]
@shaoguang123 still only shown in `goss` mode?
could you share the data and parameters for me to debug? 
And if data is large, maybe you can sample a subset that can reproduce the error.

**Comment by shaoguang123** [comment]
> @shaoguang123 still only shown in `goss` mode?
> could you share the data and parameters for me to debug?
> And if data is large, maybe you can sample a subset that can reproduce the error.

 Not only shown in 'goss' mode. When I use the default boost bype "gdbt" with the version `temp-goss,`,it is still check failed.
I'm afraid that I cannot share this data, I will try to find some public datasets to reproduct this case. if I find it, I'll send it to you.

**Comment by guolinke** [comment]
@shaoguang123 can you try the update in the temp-goss (https://github.com/microsoft/LightGBM/commit/0e32d2bc9edce0e24ca8244a2f4eb3a655bd77a2) again ?

**Comment by shaoguang123** [comment]
> @shaoguang123 can you try the update in the temp-goss ([0e32d2b](https://github.com/microsoft/LightGBM/commit/0e32d2bc9edce0e24ca8244a2f4eb3a655bd77a2)) again ?

@guolinke With the latest `temp-goss`, it seems that the problem doesn't happen anymore. Where do you think the problem might be？

**Comment by shaoguang123** [comment]
> > @shaoguang123 can you try the update in the temp-goss ([0e32d2b](https://github.com/microsoft/LightGBM/commit/0e32d2bc9edce0e24ca8244a2f4eb3a655bd77a2)) again ?
> 
> @guolinke With the latest `temp-goss`, it seems that the problem doesn't happen anymore. Where do you think the problem might be？

Sorry，it seems that some other problems happened. I'll try to do more test.

**Comment by shaoguang123** [comment]
@guolinke the temp-goss version was stuck at the 25th iteration as my learning_rate is 0.04, just like the master version before.

**Comment by shaoguang123** [comment]
The program is stuck at the code as below:
![微信图片_20200512131405](https://user-images.githubusercontent.com/7996409/81641073-bd288580-9452-11ea-8078-6bc9e5e257c5.png)

**Comment by guolinke** [comment]
@shaoguang123 
can you try the update in that branch again? I add some debug outputs, so please paste the related log.
without the reproduce code, it is hard to detect the problem, so it will be better if you can provide more information if there are problems.

**Comment by guolinke** [comment]
@shaoguang123 I think the problem is found. please try the fix in #3077

**Comment by shaoguang123** [comment]
> @shaoguang123 I think the problem is found. please try the fix in #3077

@guolinke With the version `const-hess-goss`, it seems that the convergence problem no longer happens. Where do you think the problem might be？

**Comment by guolinke** [comment]
as the fix in #3077, the (weighted) hessian is wrong for the regression problem.

**Comment by shaoguang123** [comment]
> as the fix in #3077, the (weighted) hessian is wrong for the regression problem.

Before the fix, the hessian is weighted, but the flag is_constant_hessian_  is set to True. 
And the flag is_constant_hessian_  affects the construction of histogram。
Is it means before the fix, the result under goss mode is not exactly right, whether I haven't encounter the convergence problem?

**Comment by guolinke** [comment]
yeah, the goss with regression objective is wrong before that fix.

**Comment by shaoguang123** [comment]
> yeah, the goss with regression objective is wrong before that fix.

Ok, Thank you for your answers！It quite helps me a lot!  @guolinke

**Comment by shaoguang123** [comment]
@guolinke Aonther question I want to ask is that in the code before fix, since the flag is_constant_hessian_ is set to True, when we calculate the split gain and the the output of the leaves, the hessian we used is always the first element of the hessians?

In the code before fix, if we make the `multiply` always be 1, the convergence problem will also not happen any more. Is that right?

**Comment by guolinke** [comment]
@shaoguang123 yeah, but it will be a wrong algorithm, as the distribution of data is changed during training.
